ipa-core/src/protocol/ipa_prf/shuffle/mod.rs

```diff
@@ -0,0 +1,167 @@
+    expand_array_in_place(&mut y_right, input.timestamp.right(), offset);
+
+    AdditiveShare::<YS>::new(y_left, y_right)
+}
+
+// This function converts AdditiveShare obtained from shuffle protocol to OprfReport
+pub fn shuffled_to_oprfreport<YS, BK, TV, TS>(input: &AdditiveShare<YS>) -> OprfReport<BK, TV, TS>
+where
+    YS: SharedValue + ArrayAccess<Output = Boolean> + Expand<Input = Boolean>,
+    BK: SharedValue + ArrayAccess<Output = Boolean> + Expand<Input = Boolean>,
+    TV: SharedValue + ArrayAccess<Output = Boolean> + Expand<Input = Boolean>,
+    TS: SharedValue + ArrayAccess<Output = Boolean> + Expand<Input = Boolean>,
+{
+    let match_key = extract_from_shared_array::<YS, BA64>(input, 0);
+
+    let mut offset = BA64::BITS as usize;
+
+    let is_trigger = AdditiveShare::<Boolean>::new(
+        input.left().get(offset).unwrap_or(Boolean::ZERO),
+        input.right().get(offset).unwrap_or(Boolean::ZERO),
+    );
+
+    offset += 1;
+
+    let breakdown_key = extract_from_shared_array::<YS, BK>(input, offset);
+
+    offset += BK::BITS as usize;
+    let trigger_value = extract_from_shared_array::<YS, TV>(input, offset);
+
+    offset += TV::BITS as usize;
+    let timestamp = extract_from_shared_array::<YS, TS>(input, offset);
+
+    OprfReport {
+        match_key,
+        is_trigger,
+        breakdown_key,
+        trigger_value,
+        timestamp,
+    }
+}
+
+#[cfg(all(test, unit_test))]
+pub mod tests {
+    use rand::Rng;
+
+    use crate::{
+        ff::boolean_array::{BA20, BA3, BA8},
+        protocol::ipa_prf::shuffle::shuffle_inputs,
+        test_executor::run,
+        test_fixture::{ipa::TestRawDataRecord, Reconstruct, Runner, TestWorld},
+    };
+
+    #[test]
+    fn test_shuffle_inputs() {
+        const BATCHSIZE: usize = 50;
+        run(|| async {
+            let world = TestWorld::default();
+
+            let mut rng = rand::thread_rng();
+            let mut records = Vec::new();
+
+            for _ in 0..BATCHSIZE {
+                records.push({
+                    TestRawDataRecord {
+                        timestamp: rng.gen_range(0u64..1 << 20),
+                        user_id: rng.gen::<u64>(),
+                        is_trigger_report: rng.gen::<bool>(),
+                        breakdown_key: rng.gen_range(0u32..1 << 8),
+                        trigger_value: rng.gen_range(0u32..1 << 3),
+                    }
+                });
+            }
+
+            let mut result: Vec<TestRawDataRecord> = world
+                .semi_honest(records.clone().into_iter(), |ctx, input_rows| async move {
+                    shuffle_inputs::<_, BA8, BA3, BA20>(ctx, input_rows)
+                        .await
+                        .unwrap()
+                })
+                .await
+                .reconstruct();
+            assert_ne!(result, records);
+            records.sort();
+            result.sort();
+            assert_eq!(result, records);
+        });
```

Comment:
I think that is a good test :)

Reply:
Kind of. It doesn't actually test that the inputs were shuffled though. This just ensures that the same set of elements appears after shuffling.